controllers/drain_controller.go

```diff
@@ -243,6 +293,60 @@ func (dr *DrainReconcile) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
 			"DrainController",
 			"node drain completed")
 		return ctrl.Result{}, nil
+	} else if nodeDrainAnnotation == constants.DevicePluginResetRequired {
+		// nothing to do here we need to wait for the node to move back to idle
+		if nodeStateDrainAnnotationCurrent == constants.DrainComplete {
+			reqLogger.Info("node requested a drain and nodeState is on drain completed nothing todo")
+			return ctrl.Result{}, nil
+		}
+
+		// if we are on idle state we move it to drain
+		if nodeStateDrainAnnotationCurrent == constants.DrainIdle {
+			err = utils.AnnotateObject(ctx, nodeNetworkState, constants.NodeStateDrainAnnotationCurrent, constants.Draining, dr.Client)
+			if err != nil {
+				reqLogger.Error(err, "failed to annotate node with annotation", "annotation", constants.Draining)
+				return ctrl.Result{}, err
+			}
+			return ctrl.Result{}, nil
+		}
+
+		// This cover a case where we only need to reset the device plugin
+		// for that we are going to cordon the node, so we don't get new pods allocated
+		// to the node in the time we remove the device plugin
+		err = dr.drainer.RunCordonOrUncordon(ctx, node, true)
+		if err != nil {
+			log.Log.Error(err, "failed to cordon on node")
+			return reconcile.Result{}, err
+		}
+
+		// we switch the sriov label to disable and mark the drain as completed
+		// no need to wait for the device plugin to exist here as we cordon the node,
+		// and we want to config-daemon to start the configuration in parallel of the kube-controller to remove the pod
+		// we check the device plugin was removed when the config-daemon moves is desire state to idle
+		reqLogger.Info("disable Device plugin from node")
+		err = utils.LabelNode(ctx, node.Name, constants.SriovDevicePluginEnabledLabel, constants.SriovDevicePluginEnabledLabelDisabled, dr.Client)
+		if err != nil {
+			log.Log.Error(err, "failed to label node for device plugin label",
+				"labelKey",
+				constants.SriovDevicePluginEnabledLabel,
+				"labelValue",
+				constants.SriovDevicePluginEnabledLabelDisabled)
+			return reconcile.Result{}, err
+		}
+
+		// if we manage to cordon we label the node state with drain completed and finish
+		err = utils.AnnotateObject(ctx, nodeNetworkState, constants.NodeStateDrainAnnotationCurrent, constants.DrainComplete, dr.Client)
+		if err != nil {
+			reqLogger.Error(err, "failed to annotate node with annotation", "annotation", constants.DrainComplete)
+			return ctrl.Result{}, err
+		}
+
+		reqLogger.Info("node cordoned successfully and device plugin removed")
+		dr.recorder.Event(nodeNetworkState,
+			corev1.EventTypeWarning,
+			"DrainController",
+			"node cordoned and device plugin removed completed")
+		return ctrl.Result{}, nil
```

Comment:
Any chance to code a unit test for this case? It took me a while to understand the whole controller - daemon interaction